From: Sebastian Ott <sebott@linux.vnet.ibm.com>
Subject: s390/pci: add extra padding to function measurement block
Patch-mainline: v4.6-rc5
Git-commit: 9d89d9e61d361f3adb75e1aebe4bb367faf16cfa
References: bnc#968500, LTC#139445

Description:  s390/pci: enforce fmb page boundary rule
Symptom:      Kernel panic.
Problem:      Unhandled operand exception due to a misaligned function
              measurement block.
Solution:     Enforce correct alignment.
Reproduction: PCI hotplug in a tight loop during load.

Upstream-Description:

              s390/pci: add extra padding to function measurement block

              Newer machines might use a different (larger) format for function
              measurement blocks. To ensure that we comply with the alignment
              requirement on these machines and prevent memory corruption (when
              firmware writes more data than we expect) add 16 padding bytes
              at the end of the fmb.

              Cc: stable@vger.kernel.org # v4.1+
              Signed-off-by: Sebastian Ott <sebott@linux.vnet.ibm.com>
              Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>


Signed-off-by: Sebastian Ott <sebott@linux.vnet.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 arch/s390/include/asm/pci.h |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/arch/s390/include/asm/pci.h
+++ b/arch/s390/include/asm/pci.h
@@ -50,7 +50,8 @@ struct zpci_fmb {
 	u64 rpcit_ops;
 	u64 dma_rbytes;
 	u64 dma_wbytes;
-} __packed __aligned(64);
+	u64 pad[2];
+} __packed __aligned(128);
 
 struct msi_map {
 	unsigned long irq;
